Add `#[rustc_pass_indirectly_in_non_rustic_abis]`

beetrees opened the pull request:
This PR adds an internal `#[rustc_pass_indirectly_in_non_rustic_abis]` attribute that can be applied to structs. Structs marked with this attribute will always be passed using `PassMode::Indirect { on_stack: false, .. }` when being passed by value to functions with non-Rustic calling conventions. This is needed by rust-lang/rust#141980; see that PR for further details.

cc @joshtriplett

beetrees committed:
Add `#[rustc_pass_indirectly_in_non_rustic_abis]`

RalfJung commented:
@jdonszelmann could you have a look at the attribute code here? This is my first time actually seeing the new infrastructure so I can't say if the way it is used here is correct.

RalfJung commented on compiler/rustc_target/src/callconv/riscv.rs:
```diff
@@ -293,6 +293,10 @@ fn classify_arg<'a, Ty, C>(
         return;
     }
+    if arg.layout.pass_indirectly_in_non_rustic_abis(cx) {
+        arg.make_indirect();
```
Usually loongarch and riscv do exactly the same thing. Why does this one here not update `avail_gprs`?

RalfJung commented on compiler/rustc_target/src/callconv/x86_64.rs:
```diff
@@ -186,6 +186,11 @@ where
             return;
         }
+        if is_arg && arg.layout.pass_indirectly_in_non_rustic_abis(cx) {
+            int_regs = int_regs.saturating_sub(1);
```
Very odd that this would happen here but not for unsized arguments... it's pre-existing, but can you add a FIXME? (Also for the other files with the same pattern)

RalfJung reviewed (commented):
The compiler part LGTM apart from these comments, but I can't really review these ABI adjustments.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16664096249/job/47167072264) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47167072264) [(plain)](https://github.com/rust-lang/rust/commit/d91160a2541b3ea366683df32dfbeb935002ea9d/checks/47167072264/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

SparrowLii commented:
r? @joshtriplett :)

folkertdev commented:
Besides Jubilee, who is not available, the only other person I can think of to review ABI code is 

r? @bjorn3

joshtriplett commented:
Broadly speaking, this seems like a reasonable way to tag types for use with this special calling convention.

I can't speak to the maintainability of this approach, but generally speaking, I'm not sure it's possible to handle `va_args` any *easier* than this, for any ABIs where it's a special case. I'd suggest, in general, that when you add this attribute to a *type*, there should be copious cross-references between the type, the ABI tests that confirm it's passed correctly, and the compiler sources implementing this attribute, so that people who are adding a new ABI know to carefully add tests and check the passing of `va_args`. (And, hopefully, not duplicate the weird mistakes of existing architectures.)

folkertdev commented on compiler/rustc_abi/src/layout/ty.rs:
```diff
@@ -269,6 +271,24 @@ impl<'a, Ty> TyAndLayout<'a, Ty> {
+    where
+        Ty: TyAbiInterface<'a, C> + Copy,
+    {
+        while self.is_transparent()
+            && let Some((_, field)) = self.non_1zst_field(cx)
+        {
+            self = field;
+        }
+        Ty::is_pass_indirectly_in_non_rustic_abis_flag_set(self)
```
so right now the type is only passed indirectly if the base type is passed indirectly. This just confused me for a while, I think this should also work:

```rust
#[repr(C)]
struct Vanilla(i32);

#[repr(transparent)]
#[rustc_pass_indirectly_in_non_rustic_abis]
struct Indirectly(Vanilla);
```

I.e. if any of the transparent wrappers is `rustc_pass_indirectly_in_non_rustic_abis` then the whole type should be passed indirectly.

RalfJung replied:
I think an error makes most sense -- the behavior you now implemented seems to violate the promise that repr(transparent) wrappers have the same ABI as their underlying type on *all* ABIs.

folkertdev replied:
Hmm yeah, looking at this again, given e.g. https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=7e1399219e6854e5ca5e5765cb8e6993 an error is probably better. `#[rustc_pass_indirectly_in_non_rustic_abis]` does feel morally like a repr hint.

RalfJung commented:
> Broadly speaking, this seems like a reasonable way to tag types for use with this special calling convention.
> 
> I can't speak to the maintainability of this approach, but generally speaking, I'm not sure it's possible to handle `va_args` any _easier_ than this, for any ABIs where it's a special case. I'd suggest, in general, that when you add this attribute to a _type_, there should be copious cross-references between the type, the ABI tests that confirm it's passed correctly, and the compiler sources implementing this attribute, so that people who are adding a new ABI know to carefully add tests and check the passing of `va_args`. (And, hopefully, not duplicate the weird mistakes of existing architectures.)

Correct use of the attribute depends not just on the type but also its surrounding convention -- for instance, it is already UB in C to even look at a va_args after it has been passed to another function, if I understood correctly. That's crucial because a Rust "pass_indirectly" type still passes a *copy* of the argument value to the callee, whereas in C it passes effectively an implicit reference -- if the callee mutates the argument, a C caller will see the mutations, a Rust caller will not.

RalfJung commented on compiler/rustc_target/src/callconv/amdgpu.rs:
```diff
@@ -10,11 +10,15 @@ where
     Ty: TyAbiInterface<'a, C> + Copy,
     C: HasDataLayout,
 {
+    if arg.layout.pass_indirectly_in_non_rustic_abis(cx) {
```
We have some code doing sanity checks on the ABI after it got computed. We could probably add an assertion there.

https://github.com/rust-lang/rust/blob/d71ed8d19bcfdbd6eb3bf2620071f2ab95470a57/compiler/rustc_ty_utils/src/abi.rs#L368-L372

RalfJung replied:
> How do we make sure that triggers for every ABI though? 

It's an unstable attribute, so worst case we get an ICE when building libcore. That seems fine.

folkertdev replied:
Using `const` does work, I added that and a fix for transparent wrappers ignoring the attribute at https://github.com/rust-lang/rust/compare/master...folkertdev:rust:pass-indirectly-attr-updates

@beetrees feel free to steal or chery-pick from that. I'd also happily force-push to this branch if you don't have time/interest. (this is on the critical path for c-variadics now that the error messages are in a good state, and I'd hate to waste the reviewer momentum).

folkertdev committed:
assert that `#[rustc_pass_indirectly_in_non_rustic_abis]` is respected

folkertdev committed:
additional tests for `pass_indirectly_in_non_rustic_abis`

Also emit an error when `rustc_pass_indirectly_in_non_rustic_abis` is
used in combination with `repr(transparent)`.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-20`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18265625252/job/51999536201) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51999536201) [(plain)](https://github.com/rust-lang/rust/commit/6055f0748c0fa12b7e645f39357167098eefe160/checks/51999536201/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
194                                Initialized {

213                                index: 0,
214                            },
215                            max_repr_align: None,
-                            unadjusted_abi_align: Align(8 bytes),
+                            unadjusted_abi_align: Align(4 bytes),
217                            randomization_seed: $SEED,
218                        },
219                    },

223                            arg_ext: None,
224                            pointee_size: Size(8 bytes),
225                            pointee_align: Some(
-                                Align(8 bytes),
+                                Align(4 bytes),
227                            ),
228                        },
229                        meta_attrs: None,


The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args abi/pass-indirectly-attr.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/abi/pass-indirectly-attr.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=i686-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/abi/pass-indirectly-attr" "-A" "unused" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/i686-unknown-linux-gnu/native/rust-test-helpers" "-Clinker=x86_64-linux-gnu-gcc" "-O"
stdout: none
--- stderr -------------------------------
error: fn_abi_of(extern_c) = FnAbi {
           args: [
               ArgAbi {
                   layout: TyAndLayout {
                       ty: Type,
                       layout: Layout {
---
                           unadjusted_abi_align: Align(1 bytes),
                           randomization_seed: 17313912233329922972,
                       },
                   },
                   mode: Indirect {
                       attrs: ArgAttributes {
                           regular: NoAlias | CapturesAddress | NonNull | NoUndef,
                           arg_ext: None,
                           pointee_size: Size(1 bytes),
                           pointee_align: Some(
                               Align(1 bytes),
                           ),
                       },
                       meta_attrs: None,
                       on_stack: false,
                   },
---
           can_unwind: false,
       }
  --> /checkout/tests/ui/abi/pass-indirectly-attr.rs:13:1
   |
LL | pub extern "C" fn extern_c(_: Type) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: fn_abi_of(extern_rust) = FnAbi {
           args: [
               ArgAbi {
                   layout: TyAndLayout {
                       ty: Type,
                       layout: Layout {
---
                           unadjusted_abi_align: Align(1 bytes),
                           randomization_seed: 17313912233329922972,
                       },
                   },
                   mode: Cast {
                       pad_i32: false,
                       cast: CastTarget {
                           prefix: [
                               None,
                               None,
                               None,
                               None,
                               None,
---
                               },
                               total: Size(1 bytes),
                               is_consecutive: false,
                           },
                           attrs: ArgAttributes {
                               regular: ,
                               arg_ext: None,
                               pointee_size: Size(0 bytes),
                               pointee_align: None,
                           },
                       },
                   },
               },
           ],
---
   |
LL | pub extern "Rust" fn extern_rust(_: Type) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: fn_abi_of(wrapped_transparent) = FnAbi {
           args: [
               ArgAbi {
                   layout: TyAndLayout {
                       ty: Wrapper,
                       layout: Layout {
---
                           unadjusted_abi_align: Align(4 bytes),
                           randomization_seed: 4091198607797840817,
                       },
                   },
                   mode: Indirect {
                       attrs: ArgAttributes {
                           regular: NoAlias | CapturesAddress | NonNull | NoUndef,
                           arg_ext: None,
                           pointee_size: Size(8 bytes),
                           pointee_align: Some(
                               Align(4 bytes),
                           ),
                       },
                       meta_attrs: None,
                       on_stack: false,
                   },
```

</details>

bjorn3 commented on compiler/rustc_target/src/callconv/loongarch.rs:
```diff
@@ -284,9 +284,15 @@ fn classify_arg<'a, Ty, C>(
     Ty: TyAbiInterface<'a, C> + Copy,
 {
     if !arg.layout.is_sized() {
+        // FIXME: Update avail_gprs?
```
Unsized arguments are unstable and shouldn't be used outside of the rust ABI. If you try to use them with the C ABI we are forced to invent an ABI for them. Maybe we should just error when trying to use an unsized type with a non-rust ABI?

folkertdev replied:
You get a bunch of warnings if you try, but no error https://godbolt.org/z/haPra1Ecf

```
warning: the feature `unsized_fn_params` is internal to the compiler or standard library
 --> <source>:1:12
  |
1 | #![feature(unsized_fn_params)]
  |            ^^^^^^^^^^^^^^^^^
  |
  = note: using it is strongly discouraged
  = note: `#[warn(internal_features)]` on by default

warning: `extern` fn uses type `[u8]`, which is not FFI-safe
  --> <source>:11:22
   |
11 | extern "C" fn foo(x: X) {}
   |                      ^ not FFI-safe
   |
   = help: consider using a raw pointer instead
   = note: slices have no C equivalent
   = note: `#[warn(improper_ctypes_definitions)]` on by default
```

the ABI that is used is apparently to pass a pointer and a length https://godbolt.org/z/3WWoccvr4

bjorn3 replied:
> We can't panic here, that would lead to ICEs. And we can't return an error, there's no Result.

The check can be in `adjust_for_foreign_abi`. This function doesn't return a `Result`, but the caller of the caller of this function does already return `Result<&'tcx FnAbi, &'tcx FnAbiError>`, so it should be possible to add another variant to `FnAbiError` for this.

RalfJung replied:
Currently, the one ABI check we do have is entirely separate, it's in `compiler/rustc_monomorphize/src/mono_checks/abi_check.rs`. That has the advantage of being able to point at where the ABI comes up.

It would be odd to have two entirely parallel systems for rejecting types that are invalid for an ABI.

folkertdev commented on compiler/rustc_target/src/callconv/mips.rs:
```diff
@@ -14,15 +14,22 @@ where
     let dl = cx.data_layout();
+    if arg.layout.pass_indirectly_in_non_rustic_abis(cx) {
+        arg.make_indirect();
+        *offset += dl.pointer_size();
```
Fixed.

bors commented:
:umbrella: The latest upstream changes (presumably #148356) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:pushpin: Commit 27a34ec584716006d514dcc3fbbf82ce717f47ee has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 27a34ec584716006d514dcc3fbbf82ce717f47ee -->
<!-- homu: {"type":"Approved","sha":"27a34ec584716006d514dcc3fbbf82ce717f47ee","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Zalathar commented:
Yielding to enclosing rollup.

@bors retry

folkertdev reviewed (commented):
How should I evaluate that regression? There is one minor optimization in a comment below, but beyond that I think the cost is additional queries for the attribute flag, and that is impossible to work around I think (plus should be mostly cached).

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/144529/90b65889799733f21ebdf59d96411aa531c5900a..7be6d6f2e8742cd55a2ce8b6004eb76687c46bee)*

Zalathar commented:
I thought the regression looked real, but the same benchmarks bounced back in https://github.com/rust-lang/rust/pull/148507#issuecomment-3491567371, so maybe it’s noise?

RalfJung commented on compiler/rustc_abi/src/layout/ty.rs:
```diff
@@ -269,6 +271,30 @@ impl<'a, Ty> TyAndLayout<'a, Ty> {
+    {
+        while self.is_transparent()
+            && let Some((_, field)) = self.non_1zst_field(cx)
+        {
```
I just realized that this will not do the right thing for a repr(transparent) wrapper around a 1-ZST `pass_indirectly_in_non_rustic_abis` type.

Maybe we should add an assertion against 1-ZST here or so?
